### PR TITLE
Support complex (mapping-style) PluginType.

### PR DESCRIPTION
**Pull request by dmikurube**
@muga It's WIP yet (no tests!), but can I have your impressions on this change?

### CODE REVIEW

**Comment by muga** [comment]
@dmikurube Agree. I also think that this's necessary for adding new PluginSource. 

I'm OK but, before I start reviewing the implementation, I want you to confirm that we include "Maven"PluginType in this PR? If so, we need to start talking about the spec. of the type option for MavenPluginSource here. Or split PRs later?

**Comment by dmikurube** [comment]
@muga Thanks. Agreed, we need to discuss the spec before merging this.

I've sent a design doc about Maven plugin source to you. Let's discuss based on that later.

**Comment by dmikurube** [comment]
@muga For the time being, removed the Maven support. The PR now includes only mapping-style configurations.

**Comment by muga** [comment]
@dmikurube ok, thank you for fixing. I take a look.

**Comment by dmikurube** [comment]
Jfyi: made one change. `PluginType.createFromString` is now `private`.

**Comment by dmikurube** [comment]
@muga Oops, sorry, confirmed this does not actually work for the example below while tests pass.

```
in:
  type:
    name: file
```

Will fix.

**Comment by muga** [comment]
@dmikurube good catch 👍 

I have one more question (I should've checked your design doc before but). In this PR, who can recognize (or detect) that input plugin type is default and output plugin type is maven like following? Or the detection or searching is out of scope of this PR?

```
in:
  type: file
  ... ...
out:
  type: {source: maven, name: s3, ...}
  ... ...
```

**Comment by dmikurube** [comment]
@muga Thanks for your comment. I think `PluginType` is used in `PluginSource`s to load plugins. Then, for example from #592, recognizing in `PluginSource` may be reasonable like:

https://github.com/embulk/embulk/pull/592/files#diff-b1ebb6ca84d7cd5fe5762cf0ba80ef07R68

**Comment by dmikurube** [comment]
We may need `Map<String, ToString>` or like that instead of `Map<String, String>`... `ToString` is in `org.embulk.spi.util`, though.

**Comment by dmikurube** [comment]
Ah, no, a type can have only one `@JsonCreator`. Will fix.
https://stackoverflow.com/questions/15931082/how-to-deserialize-a-class-with-overloaded-constructors-using-jsoncreator

**Comment by muga** [comment]
Agree. `PluginType` objects are used for loading and instantiating plugins in `PluginSource`. My question is who (or how) `PluginType` objects can be created in `BulkLoader` class in the case that multiple plugin types are specified in Embulk config?

**Comment by dmikurube** [comment]
Ah, got your question. `PluginType` instances are always created via Jackson's `@JsonCreator`. (There's an error there as mentioned above, though.)

Jackson + YAML maps given JSON (YAML) into `PluginType` through their `ObjectMapper` via the `@JsonCreator` method.

**Comment by muga** [comment]
@dmikurube Ah got it. Thank you for explaining. I didn't notice about it. It should work 👍

**Comment by dmikurube** [comment]
@muga Finally it recognizes the JSON below with the fix dd8329d. Can you have a look?

```
in:
  type:
    name: file
```

**Comment by dmikurube** [comment]
Added tests that actually parse configs.

File: embulk-core/src/main/java/org/embulk/exec/BulkLoader.java [hunk 1]
```diff
@@ -512,7 +512,7 @@ public void doCleanup(ConfigSource config, ResumeState resume)
     private ExecutorPlugin newExecutorPlugin(BulkLoaderTask task)
     {
         return Exec.newPlugin(ExecutorPlugin.class,
-                task.getExecConfig().get(PluginType.class, "type", new PluginType("local")));
+                task.getExecConfig().get(PluginType.class, "type", PluginType.LOCAL));
```
Comment (muga) [inline embulk-core/src/main/java/org/embulk/exec/BulkLoader.java]: Minor: Up to you. I'm not sure, we don't have major reasons so much to declare new `LOCAL` static field in `PluginType` class? `DefaultPluginType.create("local")` method call is good enough here instead?

Comment (dmikurube) [inline embulk-core/src/main/java/org/embulk/exec/BulkLoader.java]: One reason is that I wanted to keep the creator method (`JsonCreator`) `private` until we really need that `public`.

In addition, that "local" type is substantially a constant. Making the `PluginType` value constant seemed more reasonable than making the string `"local"` constant.

**Comment by dmikurube** [comment]
@muga Thanks! Will be merging after considering whether include this in releasing 0.8.24 or not.

File: embulk-core/src/main/java/org/embulk/exec/BulkLoader.java (hunk 1 above)
Comment (muga) [inline embulk-core/src/main/java/org/embulk/exec/BulkLoader.java]: Understand. thank you for explaining.

**Comment by dmikurube** [comment]
0.8.24 has been released. Merging this for 0.8.25.

**Pull request merged**